core/src/km_core_state_api.cpp

```diff
@@ -363,4 +366,28 @@ km_core_cp * km_core_state_context_debug(
   result[s.size()] = 0;
 
   return result;
-}
\ No newline at end of file
+}
+
+bool
+km_core_state_has_type(km_core_state *state, uint8_t type) {
+  return std::any_of(
+      state->actions().begin(), state->actions().end(),
+        [type](const km::core::action &a) { return a.type == type; });
+}
+
+bool
+km_core_state_should_clear_context(km_core_state *state,
+                     km_core_virtual_key vk,
+                     uint16_t modifier_state,
+                     uint8_t is_key_down,
+                     uint16_t event_flags) {
+  // if emit_keystroke is present, check if a context reset is needed
+  if (km_core_state_has_type(state, KM_CORE_IT_EMIT_KEYSTROKE)) {
+    if (km::core::vkey_to_contextreset[vk]) {
+      return true;
+    } else if (vk == KM_CORE_VKEY_BKSP && km_core_state_has_type(state, KM_CORE_IT_BACK)) {
```

Comment:
I am not sure this condition is the same as the current behaviour is it?  It does not reset the context if it is emitting a backspace and there is context to delete [to the proccessors knowledge]. It only does it if the context is empty. Which I know theoretical is then a no-op.  I think you could use `code_points_to_delete` or check the context as we haven't applied the actions yet at this point. 
Current behaviour is at https://github.com/keymanapp/keyman/blob/7dcbcf35f1e6ea88d7a1a7e70087c860d6c5efe4/core/src/kmx/kmx_processevent.cpp#L278

Reply:
Does it work out the same because this action only applies if `KM_CORE_IT_EMIT_KEYSTROKE` is true?

Reply:
No this part is wip.

Reply:
can't use `code_points_to_delete` because that struct has been changed into an action array already by `state->set_actions()`.

Reply:
take a look now

Reply:
> can't use `code_points_to_delete` because that struct has been changed into an action array already by `state->set_actions()`.

Ah yes I see you are actually working on the action que and not action struct. It is getting a little confusing now. 
from LDML processor it is struct -> queue then core again -> struct for the engines.  This PR could possibly just work of the struct as the queue will eventually be deprecated.